File: src/lib/tiposkripto/src/BaseSetup.ts
import type { TestTypeParams_any } from "./CoreTypes.js";
import type { ITestArtifactory, ITestResourceConfiguration } from "./types.js";

/**
 * BaseSetup is the internal unified base class for all setup phases.
 * It covers BDD's Given, AAA's Arrange, and TDT's Map.
 * This class is not exposed to users - use BaseGiven, BaseValue, or BaseDescribe instead.
 */
export abstract class BaseSetup<I extends TestTypeParams_any> {
  features: string[];
  actions: any[];
  checks: any[];
  error: Error;
  fail: any;
  store: I["istore"];
  recommendedFsPath: string = "";
  setupCB: I["given"];
  initialValues: any;
  key: string;
  failed: boolean;
  artifacts: string[] = [];
  fails: number = 0;
  status: boolean | undefined;

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path,
        )}`,
      );
    }
    const normalizedPath = path.replace(/\\/g, "/");
    this.artifacts.push(normalizedPath);
  }

  constructor(
    features: string[],
    actions: any[],
    checks: any[],
    setupCB: I["given"],
    initialValues: any,
  ) {
    this.features = features;
    this.actions = actions;
    this.checks = checks;
    this.setupCB = setupCB;
    this.initialValues = initialValues;
    this.fails = 0;
    this.failed = false;
    this.error = null as any;
    this.store = null;
    this.key = "";
    this.status = undefined;
  }

  toObj() {
    return {
      key: this.key,
      actions: (this.actions || []).map((a) => {
        if (a && a.toObj) return a.toObj();
        console.error("Action step is not as expected!", JSON.stringify(a));
        return {};
      }),
      checks: (this.checks || []).map((c) => (c && c.toObj ? c.toObj() : {})),
      error: this.error ? [this.error, this.error.stack] : null,
      failed: this.failed,
      features: this.features || [],
      artifacts: this.artifacts,
      status: this.status,
    };
  }

  abstract setupThat(
    subject: I["isubject"],
    testResourceConfiguration: ITestResourceConfiguration,
    artifactory: ITestArtifactory,
    setupCB: I["given"],
    initialValues: any,
  ): Promise<I["istore"]>;

  async afterEach(
    store: I["istore"],
    key: string,
    artifactory: ITestArtifactory,
  ): Promise<I["istore"]> {
    return store;
  }

  async setup(
    subject: I["isubject"],
    key: string,
    testResourceConfiguration: ITestResourceConfiguration,
    tester: (t: Awaited<I["then"]> | undefined) => boolean,
    artifactory?: ITestArtifactory,
    suiteNdx?: number,
  ) {
    this.key = key;
    this.fails = 0;

    const actualArtifactory =
      artifactory || ((fPath: string, value: unknown) => { });
    const setupArtifactory = (fPath: string, value: unknown) =>
      actualArtifactory(`setup-${key}/${fPath}`, value);

    try {
      this.store = await this.setupThat(
        subject,
        testResourceConfiguration,
        setupArtifactory,
        this.setupCB,
        this.initialValues,
      );
      this.status = true;
    } catch (e: any) {
      this.status = false;
      this.failed = true;
      this.fails++;
      this.error = e;
      return this.store;
    }

    try {
      // Process actions
      for (const [actionNdx, actionStep] of (this.actions || []).entries()) {
        try {
          // Create artifactory for action context
          const actionArtifactory = this.createArtifactoryForAction(
            key,
            actionNdx,
            suiteNdx,
          );
          this.store = await actionStep.test(
            this.store,
            testResourceConfiguration,
            actionArtifactory,
          );
        } catch (e: any) {
          this.failed = true;
          this.fails++;
          this.error = e;
        }
      }

      // Process checks
      for (const [checkNdx, checkStep] of this.checks.entries()) {
        try {
          const filepath =
            suiteNdx !== undefined
              ? `suite-${suiteNdx}/setup-${key}/check-${checkNdx}`
              : `setup-${key}/check-${checkNdx}`;
          // Create artifactory for check context
          const checkArtifactory = this.createArtifactoryForCheck(
            key,
            checkNdx,
            suiteNdx,
          );
          const t = await checkStep.test(
            this.store,
            testResourceConfiguration,
            filepath,
            checkArtifactory,
          );
          tester(t);
        } catch (e: any) {
          this.failed = true;
          this.fails++;
          this.error = e;
        }
      }
    } catch (e: any) {
      this.error = e;
      this.failed = true;
      this.fails++;
    } finally {
      try {
        await this.afterEach(this.store, this.key, setupArtifactory);
      } catch (e: any) {
        this.failed = true;
        this.fails++;
        this.error = e;
      }
    }

    return this.store;
  }

  private createArtifactoryForAction(
    key: string,
    actionIndex: number,
    suiteNdx?: number,
  ) {
    const self = this as any;
    if (self._parent && self._parent.createArtifactory) {
      return self._parent.createArtifactory({
        givenKey: key,
        whenIndex: actionIndex,
        suiteIndex: suiteNdx,
      });
    }
    // Fallback to logging
    return {
      writeFileSync: (filename: string, payload: string) => {
        let path = "";
        if (suiteNdx !== undefined) {
          path += `suite-${suiteNdx}/`;
        }
        path += `setup-${key}/`;
        path += `action-${actionIndex} ${filename}`;
        console.log(`[Artifactory] Would write to: ${path}`);
        console.log(`[Artifactory] Content: ${payload.substring(0, 100)}...`);
      },
      screenshot: (filename: string, payload?: string) => {
        console.log(`[Artifactory] Would take screenshot: ${filename}`);
      },
    };
  }

  private createArtifactoryForCheck(
    key: string,
    checkIndex: number,
    suiteNdx?: number,
  ) {
    const self = this as any;
    if (self._parent && self._parent.createArtifactory) {
      return self._parent.createArtifactory({
        givenKey: key,
        thenIndex: checkIndex,
        suiteIndex: suiteNdx,
      });
    }
    // Fallback to logging
    return {
      writeFileSync: (filename: string, payload: string) => {
        let path = "";
        if (suiteNdx !== undefined) {
          path += `suite-${suiteNdx}/`;
        }
        path += `setup-${key}/`;
        path += `check-${checkIndex} ${filename}`;
        console.log(`[Artifactory] Would write to: ${path}`);
        console.log(`[Artifactory] Content: ${payload.substring(0, 100)}...`);
      },
      screenshot: (filename: string, payload?: string) => {
        console.log(`[Artifactory] Would take screenshot: ${filename}`);
      },
    };
  }
}

export type ISetups<I extends TestTypeParams_any> = Record<
  string,
  BaseSetup<I>
>;

File: src/lib/rubeno/lib/base_setup.rb
module Rubeno
  # BaseSetup is the unified base class for all setup phases.
  # It covers BDD's Given, AAA's Arrange, and TDT's Map.
  class BaseSetup
    attr_accessor :features, :actions, :checks, :error, :fail, :store, :recommended_fs_path,
                  :setup_cb, :initial_values, :key, :failed, :artifacts, :status, :fails, :_parent
    
    def initialize(features, actions, checks, setup_cb, initial_values)
      @features = features
      @actions = actions
      @checks = checks
      @setup_cb = setup_cb
      @initial_values = initial_values
      @artifacts = []
      @fails = 0
      @failed = false
      @error = nil
      @store = nil
      @key = ""
      @status = nil
      @_parent = nil
    end
    
    def add_artifact(path)
      if !path.is_a?(String)
        raise "[ARTIFACT ERROR] Expected string, got #{path.class}: #{path.inspect}"
      end
      normalized_path = path.gsub('\\', '/')
      @artifacts << normalized_path
    end
    
    def to_obj
      {
        key: @key,
        actions: (@actions || []).map { |a| a.respond_to?(:to_obj) ? a.to_obj : {} },
        checks: (@checks || []).map { |c| c.respond_to?(:to_obj) ? c.to_obj : {} },
        error: @error ? [@error, @error.backtrace] : nil,
        failed: @failed,
        features: @features || [],
        artifacts: @artifacts,
        status: @status
      }
    end
    
    def setup_that(subject, test_resource_configuration, artifactory, setup_cb, initial_values)
      # This should be implemented by subclasses
      raise NotImplementedError, "Subclasses must implement setup_that"
    end
    
    def after_each(store, key, artifactory)
      store
    end
    
    def setup(subject, key, test_resource_configuration, tester, artifactory = nil, suite_ndx = nil)
      @key = key
      @fails = 0
      
      actual_artifactory = artifactory
      unless actual_artifactory
        # Try to get artifactory from parent
        if self.respond_to?(:_parent) && @_parent && @_parent.respond_to?(:create_artifactory)
          actual_artifactory = @_parent.create_artifactory(given_key: key, suite_index: suite_ndx)
        else
          # Create a default artifactory
          actual_artifactory = {
            write_file_sync: ->(filename, payload) {
              puts "[Artifactory] Would write to: #{filename}"
            },
            screenshot: ->(filename, payload = nil) {
              puts "[Artifactory] Would take screenshot: #{filename}"
            }
          }
        end
      end
      
      begin
        @store = setup_that(
          subject,
          test_resource_configuration,
          actual_artifactory,
          @setup_cb,
          @initial_values
        )
        @status = true
      rescue => e
        @status = false
        @failed = true
        @fails += 1
        @error = e
        return @store
      end
      
      begin
        # Process actions
        (@actions || []).each_with_index do |action_step, action_ndx|
          begin
            # Create artifactory for action context
            action_artifactory = create_artifactory_for_action(key, action_ndx, suite_ndx, actual_artifactory)
            @store = action_step.test(
              @store,
              test_resource_configuration,
              action_artifactory
            )
          rescue => e
            @failed = true
            @fails += 1
            @error = e
          end
        end
        
        # Process checks
        @checks.each_with_index do |check_step, check_ndx|
          begin
            # Create artifactory for check context
            check_artifactory = create_artifactory_for_check(key, check_ndx, suite_ndx, actual_artifactory)
            filepath = suite_ndx ? "suite-#{suite_ndx}/setup-#{key}/check-#{check_ndx}" : 
                                   "setup-#{key}/check-#{check_ndx}"
            t = check_step.test(
              @store,
              test_resource_configuration,
              filepath,
              check_artifactory
            )
            tester.call(t)
          rescue => e
            @failed = true
            @fails += 1
            @error = e
          end
        end
      rescue => e
        @error = e
        @failed = true
        @fails += 1
      ensure
        begin
          after_each(@store, @key, actual_artifactory)
        rescue => e
          @failed = true
          @fails += 1
        end
      end
      
      @store
    end
    
    private
    
    def create_artifactory_for_action(key, action_index, suite_ndx, base_artifactory)
      # Try to get artifactory from parent
      if self.respond_to?(:_parent) && @_parent && @_parent.respond_to?(:create_artifactory)
        return @_parent.create_artifactory(given_key: key, when_index: action_index, suite_index: suite_ndx)
      end
      
      # Fallback to base artifactory
      base_artifactory
    end
    
    def create_artifactory_for_check(key, check_index, suite_ndx, base_artifactory)
      # Try to get artifactory from parent
      if self.respond_to?(:_parent) && @_parent && @_parent.respond_to?(:create_artifactory)
        return @_parent.create_artifactory(given_key: key, then_index: check_index, suite_index: suite_ndx)
      end
      
      # Fallback to base artifactory
      base_artifactory
    end
  end
end
